src/Compilers/Core/Portable/Compilation/DeterministicKeyBuilder.cs

```diff
@@ -0,0 +1,478 @@
+            writer.Write("outputNameOverride", options.OutputNameOverride);
+            writer.Write("pdbFilePath", options.PdbFilePath);
+            writer.Write("pdbChecksumAlgorithm", options.PdbChecksumAlgorithm.Name);
+            writer.Write("runtimeMetadataVersion", options.RuntimeMetadataVersion);
+            writer.Write("defaultSourceFileEncoding", options.DefaultSourceFileEncoding?.CodePage);
+            writer.Write("fallbackSourceFileEncoding", options.FallbackSourceFileEncoding?.CodePage);
+
+            writer.WriteObjectEnd();
+
+            static void writeSubsystemVersion(JsonWriter writer, SubsystemVersion version)
+            {
+                writer.WriteKey("subsystemVersion");
+                writer.WriteObjectStart();
+                writer.Write("major", version.Major);
+                writer.Write("minor", version.Minor);
+                writer.WriteObjectEnd();
+            }
+        }
+
+        private void WriteCompilationOptions(JsonWriter writer, CompilationOptions options)
+        {
+            writer.WriteObjectStart();
+            WriteCompilationOptionsCore(writer, options);
+            writer.WriteObjectEnd();
+        }
+
+        protected virtual void WriteCompilationOptionsCore(JsonWriter writer, CompilationOptions options)
+        {
+            // CompilationOption values
+            writer.Write("outputKind", options.OutputKind);
+            writer.Write("moduleName", options.ModuleName);
+            writer.Write("scriptClassName", options.ScriptClassName);
+            writer.Write("mainTypeName", options.MainTypeName);
+            WriteByteArrayValue(writer, "cryptoPublicKey", options.CryptoPublicKey);
+            writer.Write("cryptoKeyFile", options.CryptoKeyFile);
+            writer.Write("delaySign", options.DelaySign);
+            writer.Write("publicSign", options.PublicSign);
+            writer.Write("checkOverflow", options.CheckOverflow);
+            writer.Write("platform", options.Platform);
+            writer.Write("optimizationLevel", options.OptimizationLevel);
+            writer.Write("generalDiagnosticOption", options.GeneralDiagnosticOption);
+            writer.Write("warningLevel", options.WarningLevel);
+            writer.Write("deterministic", options.Deterministic);
+            writer.Write("debugPlusMode", options.DebugPlusMode);
+            writer.Write("referencesSupersedeLowerVersions", options.ReferencesSupersedeLowerVersions);
+            writer.Write("reportSuppressedDiagnostics", options.ReportSuppressedDiagnostics);
+            writer.Write("nullableContextOptions", options.NullableContextOptions);
+
+            writer.WriteKey("specificDiagnosticOptions");
+            writer.WriteArrayStart();
+            foreach (var kvp in options.SpecificDiagnosticOptions)
+            {
+                writer.WriteObjectStart();
+                writer.Write(kvp.Key, kvp.Value);
+                writer.WriteObjectEnd();
```

Comment:
this enumerates a dictionary.  that may have different results across rules of hte same program as i don't believe .net guarantees deterministic order.  Consider sorting things like this to ensure we're ok.